### PR TITLE
feat: add Domino drafter training backend

### PR DESCRIPTION
**Pull request by khazic**
## What

Adds a Domino drafter training backend to the SpeCo overlay (part of #16). Select it for drafter training with:

```
actor_rollout_ref.rollout.drafter.speculative_algorithm=DOMINO
```

## Design

Domino is a DFlash variant, so `DominoTrainerBackend` subclasses `DFlashTrainerBackend` exactly like DSpark does; only `build_model` and the training forward differ. The DFlash parallel block backbone drafts a whole block in one non-causal forward, so each predicted position is blind to the block's earlier (drafted) tokens. Domino adds a causal correction head that fixes this:

- `prefix_gru`: a single-layer GRU over the block's previous-token embeddings, producing a causal state per block position.
- `embed_proj`: a low-rank MLP over `[backbone hidden | GRU state]` emitting a full-vocabulary logit delta added to the parallel base logits.

Training jointly supervises the refined (final) and backbone-only (base) logits with a base-anchor curriculum `loss = (1 - lambda) * final + lambda * base`, where `lambda` decays from its start value to 0 so early steps keep the parallel backbone strong and later steps let the correction head take over.

The shifted-label alignment (target `x[a+1:a+1+block]`, prev `[x[a], labels[:-1]]`, every position supervised) reuses the DSpark alignment, which equals the AutoModel reference's `shift_label` Domino path. Domino reuses the DFlash block-drafter plumbing (anchor sampling, noise block, block attention mask, `dflash_aux` hidden layout) and is wired through the worker dispatch, `base_trainer` block-drafter gates, auto config routing, old-logprob aux layers, and config keys (`domino_*`, falling back to `dflash_*`).

## Serving

Domino is not an engine-level speculative algorithm. Engines expose it as a `projector_type` sub-mode of DFlash: the method stays `dflash`, and the causal correction head (`prefix_gru` + `embed_proj`) is enabled from the checkpoint's `dflash_config.projector_type="domino"`. `DOMINO` is therefore never a valid engine algorithm string, so both the vLLM and the SGLang config builders raise with a clear message directing the rollout/serve path to `DFLASH`, which keeps the Domino correction head active on engines that carry the projector.

Engine-side support is in flight and not merged yet (vllm-project/vllm#48241 for vLLM, sgl-project/sglang#31328 for SGLang), so a Domino checkpoint cannot be served on a released engine today. This PR only adds the training backend plus the two guardrails; no serving claim is made here.

## Not a duplicate

Checked the open issues/PRs (`gh pr list`): no open PR touches Domino. #13 (Separate Draft Model Training) and #10 (draft weight loading) are unrelated.

## Testing

A human reviewed every changed line and ran the tests below.

CPU contract tests (`tests/integration/test_domino_backend_contract.py`), 7 passed: projector-head shapes, the lambda-base curriculum schedule, block-drafter classification, auto-config routing for `DominoDraftModel`, old-logprob DFlash-aux routing, and the DOMINO serve guardrail on both the vLLM and the SGLang config builders. The full suite is `149 passed, 2 skipped, 2 failed`; both failures pre-exist on `main` (a `tests/heavy` runner-env check and a stale metrics assertion) and are unrelated to this change.

```
pytest tests/integration/test_domino_backend_contract.py -q
# 7 passed
```

GPU hardware smoke (`ci/domino_gpu_smoke.py`) on a real target (Qwen3-4B, 1x A100): it collects the real DFlash-style multi-layer context hidden states from the frozen target, builds the Domino draft via `DominoTrainerBackend.build_model`, and runs 120 optimizer steps through `compute_loss` (the block-drafter forward with the GRU correction head and dual-logit curriculum). The draft is cold-started, so the signal is convergence: both the final (Domino-refined) and base (backbone-only) CE fall and accuracy rises. Crucially, the **final loss stays below the base loss** throughout, i.e. the causal correction head genuinely improves on the parallel backbone, and `lambda_base` decays from ~1 to 0 as designed.

```
python ci/domino_gpu_smoke.py --target /path/to/Qwen3-4B --steps 120 --lambda-decay-steps 60
```

<details>
<summary>Domino GPU training log (Qwen3-4B target, cold start, 120 steps)</summary>

```
[smoke] loading target /path/to/Qwen3-4B
[smoke] context layers=[1, 9, 17, 25, 33] (of 36)
[smoke] batch seq_len=91 hidden=12800
[smoke] block_size=8 gru=1024 emb_dim=256 trainable_params=174,696,608
[smoke] step   0  final_loss=11.9464  base_loss=11.9907  final_acc=0.0000  base_acc=0.0000  lambda_base=0.983
[smoke] step  10  final_loss=4.9976  base_loss=5.1292  final_acc=0.0679  base_acc=0.0679  lambda_base=0.817
[smoke] step  20  final_loss=2.2520  base_loss=2.6272  final_acc=0.4090  base_acc=0.4220  lambda_base=0.650
[smoke] step  30  final_loss=0.5375  base_loss=0.9535  final_acc=0.7991  base_acc=0.7991  lambda_base=0.483
[smoke] step  40  final_loss=0.0794  base_loss=0.3261  final_acc=0.9899  base_acc=0.9393  lambda_base=0.317
[smoke] step  50  final_loss=0.0340  base_loss=0.1515  final_acc=0.9971  base_acc=0.9928  lambda_base=0.150
[smoke] step  60  final_loss=0.0240  base_loss=0.1278  final_acc=1.0000  base_acc=1.0000  lambda_base=0.000
[smoke] step  70  final_loss=0.0232  base_loss=0.1251  final_acc=1.0000  base_acc=1.0000  lambda_base=0.000
[smoke] step  80  final_loss=0.0232  base_loss=0.1249  final_acc=1.0000  base_acc=1.0000  lambda_base=0.000
[smoke] step  90  final_loss=0.0227  base_loss=0.1243  final_acc=1.0000  base_acc=1.0000  lambda_base=0.000
[smoke] step 100  final_loss=0.0227  base_loss=0.1242  final_acc=1.0000  base_acc=1.0000  lambda_base=0.000
[smoke] step 110  final_loss=0.0227  base_loss=0.1239  final_acc=1.0000  base_acc=1.0000  lambda_base=0.000
[smoke] step 119  final_loss=0.0226  base_loss=0.1239  final_acc=1.0000  base_acc=1.0000  lambda_base=0.000
[smoke] DONE  final_loss 11.9464->0.0226  base_loss 11.9907->0.1239  final_acc 0.0000->1.0000
```

</details>

## Known follow-ups

- Domino trains a `full_vocab` CE head; restricted/sampled-CE (as DFlash/DSpark support) would need to index the `embed_proj` output rows and is left as a follow-up.
- Serving is gated on the engine PRs above landing. The checkpoint this backend writes matches their config contract on paper (`projector_type`, `gru_hidden_dim`, `pure_draft_prefix_len`, `shift_label`, `emb_dim`, and the `prefix_gru.*` / `embed_proj.*` weight names), but an end-to-end load of a SpeCo-trained Domino checkpoint on those branches has not been validated yet.

### CODE REVIEW

**Comment by khazic** [comment]
Thanks for the review. Addressed the findings:

**`spec_model_path=None` guard (high):** Fixed in b23c03c. `config_path` now falls back to `None`, and the existing `if config_path and os.path.exists(...)` / short-circuited checks route a from-scratch run to the fallback config instead of crashing.

**`_forward_count` as a registered buffer (medium x2):** Not applied, because in this repo the wrapper's counter is never checkpointed or restored, so a buffer would not survive resume:

- Hot publish (`_get_trainable_state_dict`) skips non-floating tensors, so a `long` buffer is excluded anyway.
- Offline export (`_get_pretrained_export_state_dict`) strips to `draft_model.`-prefixed names only; `_forward_count` lives on the `DominoTrainingModel` wrapper, not under `draft_model.`, so it is dropped.
- Resume goes through `build_model` -> `_load_draft_checkpoint`, which loads only into `draft_model`; the wrapper is reconstructed fresh either way.

Drafter resume reloads the HF drafter weights with a fresh optimizer and schedule, so the per-run counter restarting at 0 is consistent with the rest of the training state, not a desync. Keeping it a plain `int` avoids a per-step device sync with no behavioral gain.

File: verl_speco/backends/domino_trainer_backend.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+            base_correct = torch.zeros_like(flat_weights, dtype=torch.bool)
+            top1_correct = torch.zeros((), dtype=torch.float32, device=device)
+            top5_correct = torch.zeros((), dtype=torch.float32, device=device)
+            quality_token_count = torch.zeros((), dtype=torch.float32, device=device)
+            if active_final_pred is not None and active_targets.numel() > 0:
+                active_correct = active_final_pred.eq(active_targets)
+                correct[active_mask] = active_correct
+                base_correct[active_mask] = active_base_pred.eq(active_targets)
+                top1_correct = active_correct.float().sum()
+                quality_token_count = active_targets.new_tensor(float(active_targets.numel()), dtype=torch.float32)
```
[the remark this answers is not in the document]
Comment (khazic) [reply verl_speco/backends/domino_trainer_backend.py:271]: Good catch, fixed in e2ea1a4. DFlash and DSpark both reduce this counter and Domino did not, so `top5_acc` (derived in `base_trainer`) was pinned at 0.

I took a different shape than the suggestion: mirrored the `argmax` idiom already in this block (topk over `base_logits`, suffix rows overwritten from `active_final_logits`) so no second `[num_active, vocab]` tensor is materialized, which this forward deliberately avoids. Also guarded with `topk = min(5, vocab)` like DSpark, since the suggestion hardcodes `topk(5)` and would crash on the small-vocab configs the contract tests use. Added a CPU regression test that fails without the fix (top5=0 vs top1=2).

File: verl_speco/backends/domino_trainer_backend.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+        if value is not None:
+            return value
+        return training_cfg.get(dflash_key, default)
+
+    def _normalize_dflash_config(self, drafter_config, target_hf_config, normalized_state, spec_model_path):
+        training_cfg = self.config.rollout.drafter.training
+        if training_cfg.get("domino_num_target_layers", None) is not None:
+            if getattr(drafter_config, "num_context_layers", None) is None:
+                drafter_config.num_context_layers = int(training_cfg["domino_num_target_layers"])
+        return super()._normalize_dflash_config(drafter_config, target_hf_config, normalized_state, spec_model_path)
```
[the remark this answers is not in the document]
Comment (khazic) [reply verl_speco/backends/domino_trainer_backend.py:327]: The divergence is real, but it is not introduced here and not Domino-specific: `dspark_trainer_backend.py:393` is the same code with the same `is None` guard, and `_dflash_num_context_layers` gives `dspark_num_target_layers` the same priority over the checkpoint value. Domino mirrors DSpark verbatim by design.

Adding the raise only to Domino would make the two backends behave differently on the same misconfiguration. Worth fixing, but as one change covering DSpark and Domino together rather than inside this PR.

File: verl_speco/backends/domino_trainer_backend.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+        self.target_lm_head = self._build_target_lm_head(target_model_path, target_hf_config)
+        training_cfg = self.config.rollout.drafter.training
+        return DominoTrainingModel(
+            draft_model=draft_model,
+            block_size=int(training_cfg.get("domino_block_size", getattr(drafter_config, "block_size", 16))),
+            num_anchors=int(training_cfg.get("domino_num_anchors", getattr(drafter_config, "num_anchors", 512))),
+            loss_decay_gamma=float(training_cfg.get("domino_loss_decay_gamma", getattr(drafter_config, "loss_decay_gamma", 7.0))),
+            pure_draft_prefix_len=int(training_cfg.get("domino_pure_draft_prefix_len", getattr(drafter_config, "pure_draft_prefix_len", 1))),
+            lambda_base_start=float(training_cfg.get("domino_lambda_base_start", getattr(drafter_config, "lambda_base_start", 1.0))),
+            lambda_base_decay_steps=int(training_cfg.get("domino_lambda_base_decay_steps", getattr(drafter_config, "lambda_base_decay_steps", 2000))),
```
[the remark this answers is not in the document]
Comment (khazic) [reply verl_speco/backends/domino_trainer_backend.py:405]: Same reasoning: `int(training_cfg.get(key, default))` is the pattern DSpark (`435-442`, `477-478`) and DFlash (`692-698`) already use, and the shipped `speco_base.yaml` sets all six of these keys to non-null values, so this only fires if someone explicitly writes `domino_block_size: null`. `dspark_block_size: null` has the identical hazard today.

Hardening only Domino would make it inconsistent with its siblings, so this belongs in a repo-wide change if we want it.
